historical_data_api_client/cli.py

```diff
@@ -0,0 +1,110 @@
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import json
+import logging
+import os
+from datetime import datetime
+from logging import getLogger
+
+import click
+from click.types import Choice
+from click_default_group import DefaultGroup
+from colorlog.colorlog import ColoredFormatter
+
+from d3a_api_client.utils import get_simulation_id_from_config_uuid, get_auth_headers_for_requests
+from historical_data_api_client.get_historical_data import get_historical_device_profiles
+from historical_data_api_client.utils import validate_correct_datetime_format, DATE_TIME_FORMAT
+
+log = getLogger(__name__)
+
+
+@click.group(name='d3a-api-client', cls=DefaultGroup, default='run', default_if_no_args=True,
+             context_settings={'max_content_width': 120})
+@click.option('-l', '--log-level', type=Choice(list(logging._nameToLevel.keys())), default='INFO',
+              show_default=True, help="Log level")
+def main(log_level):
+    handler = logging.StreamHandler()
+    handler.setLevel(log_level)
+    handler.setFormatter(
+        ColoredFormatter(
+            "%(log_color)s%(asctime)s.%(msecs)03d %(levelname)-8s (%(lineno)4d) %(name)-30s: "
+            "%(message)s%(reset)s",
+            datefmt="%H:%M:%S"
+        )
+    )
+    root_logger = logging.getLogger()
+    root_logger.setLevel(log_level)
+    root_logger.addHandler(handler)
+
+
+@main.command()
+@click.option('-u', '--username', type=str, help="D3A username", required=None)
+@click.option('-p', '--password', type=str, help="D3A password", required=None)
+@click.option('-c', '--configuration-id', type=str, help="configuration uuid", required=True)
+@click.option('-a', '--area-uuid', type=str, help="area uuid", required=True)
+@click.option('-d', '--domain-name', default="https://d3aweb.gridsingularity.com",
+              type=str, help="D3A domain URL")
+@click.option('-s', '--start-time', type=str, default=None,
+              help="start time (Format(YYYY-MM-DDThh:mm)")
+@click.option('-e', '--end-time', type=str, default=None,
+              help="end time (Format(YYYY-MM-DDThh:mm)")
+@click.option('-t', '--time-resolution-hours', type=int, default=1,
+              help="time resolution in hours")
+@click.option('-r', '--export-file-path', type=str, default=None,
+              help="path to export json file")
+def run(username, password, domain_name, configuration_id, area_uuid, start_time, end_time,
+        time_resolution_hours, export_file_path):
+    try:
+        if username is None:
+            if "API_CLIENT_USERNAME" not in os.environ:
+                raise Exception("Please provide a username either via cli or env variable")
```

Comment:
Nit: I would suggest using a more specific exception. I'm not sure whether this would be the best way to go, but we could also re-use one of `click`'s exceptions ([see docs](https://click.palletsprojects.com/en/8.0.x/api/#exceptions)).

Reply:
The same comment applies to the other raised Exceptions

Reply:
I know that e.g. pylint does not like the broad `Exception`. Even over the years I haven't understood the reason behind it. What does the user or maintainer gain from getting this exception:
```bash
ClickException : Please provide a username either via cli or env variable
```
or
```bash
HistoricalDataAPIException : Please provide a username either via cli or env variable
```
in comparison to this Exception:
``` bash
Exception: Please provide a username either via cli or env variable
```
Would be great if you could pitch this to me, i havent found something helpful in the internet.

Reply:
There's some examplation [here](https://realpython.com/the-most-diabolical-python-antipattern/).

The first example that comes to my mind. Suppose I am an external developer and I want to integrate the `d3a-api-client` in my own project. I might have some code like:
```python
from d3a_api_client import cli

def some_complex_function():
    data = cli.run(...)
    preprocess_data(data)

def main():
    some_complex_function()    
```
Suppose I get an exception while running `main()`. If the exception is related to the `d3a_api_client` code, I want to retry with the connection. If the exception is raised by my own preprocessing function, I want to stop and gracefully exit. 
How can I distinguish between the two cases? If the client gives me a proper exception, that's simple:
```python
def main():
    try:
        some_complex_function()
    except D3AApiClientException as ex:
        # The exception comes from the client! Try reconnecting
    except MyOwnPreprocessingException as ex:
        # The exception comes from my own code! Exit or do other stuff
```
Also another example: suppose I want to check my production logs and filter them to only find the ones related to the d3a-api-client. If I have a specific exception, I can just search for it. Searching for the error message isn't as reliable.

In general, proper exception handling allows having a better understanding of what happens.